lib/Dialect/FIRRTL/Transforms/SpecializeLayers.cpp

```diff
@@ -0,0 +1,725 @@
+
+    // Get the specialization for the current layerblock, not taking into
+    // account if an outer layerblock was specialized.  We should not have
+    // recursed inside of the disabled layerblock anyways, as it just gets
+    // erased.
+    auto specialization = getSpecialization(oldLayerRef);
+
+    // We are not specializing this layerblock.
+    if (!specialization) {
+      //  We must update the name of this layerblock to reflect
+      //  specializations of any outer layers.
+      auto newLayerRef = specializeLayerRef(oldLayerRef).getValue();
+      if (oldLayerRef != newLayerRef)
+        layerBlock.setLayerNameAttr(newLayerRef);
+      // Specialize inner operations, but keep them in their original
+      // location.
+      auto *block = layerBlock.getBody();
+      auto bodyIP = InsertionPoint::atBlockEnd(block);
+      specializeBlock(block, bodyIP, removedSyms);
+      insertionPoint.moveOpBefore(layerBlock);
+      return;
+    }
+
+    // We are enabling this layer, and all contents of this layer need to be
+    // moved (inlined) to the insertion point.
+    if (*specialization == Specialization::Enable) {
+      // Move all contents to the insertion point and specialize them.
+      specializeBlock(layerBlock.getBody(), insertionPoint, removedSyms);
+      // Erase the now empty layerblock.
+      layerBlock->erase();
+      return;
+    }
+
+    // We are disabling this layerblock, so we can just erase the layerblock. We
+    // need to record all the objects with symbols which are being deleted.
+    auto moduleName = layerBlock->getParentOfType<FModuleOp>().getNameAttr();
+    layerBlock.walk([&](hw::InnerSymbolOpInterface op) {
+      if (auto innerSym = op.getInnerSymAttr())
+        recordRemovedInnerSym(removedSyms, moduleName, innerSym);
+    });
+    layerBlock->erase();
+  }
+
+  void specializeOp(WhenOp when, InsertionPoint &insertionPoint,
+                    DenseSet<Attribute> &removedSyms) {
+    // We need to specialize both arms of the when, but the inner ops should
+    // be left where they are (and not inlined to the insertion point).
+    auto *thenBlock = &when.getThenBlock();
+    auto thenIP = InsertionPoint::atBlockEnd(thenBlock);
+    specializeBlock(thenBlock, thenIP, removedSyms);
+    if (auto *elseBlock = &when.getElseBlock()) {
+      auto elseIP = InsertionPoint::atBlockEnd(elseBlock);
+      specializeBlock(elseBlock, elseIP, removedSyms);
+    }
+    insertionPoint.moveOpBefore(when);
+  }
+
+  void specializeOp(MatchOp match, InsertionPoint &insertionPoint,
+                    DenseSet<Attribute> &removedSyms) {
+    for (size_t i = 0, e = match.getNumRegions(); i < e; ++i) {
+      auto *caseBlock = &match.getRegion(i).front();
+      auto caseIP = InsertionPoint::atBlockEnd(caseBlock);
+      specializeBlock(caseBlock, caseIP, removedSyms);
+    }
+    insertionPoint.moveOpBefore(match);
+  }
+
+  void specializeOp(InstanceOp instance, InsertionPoint &insertionPoint,
```

Comment:
```suggestion
  void specializeOp(InstanceOp instance, InsertionPoint &insertionPoint,
```

Can these re-use the same implementation? Not sure if there's an interface they share, but maybe if templated? Your call if you'd prefer it this way :+1:.  Looking around, less sure but leaving for your consideration.